docs: Attestation protocol updates

sameo opened the pull request:
Signed-off-by: Samuel Ortiz <sameo@rivosinc.com>

sameo commented:
cc @slp

jialez0 reviewed (commented):
@sameo Thanks! A few comments...

sameo commented on docs/kbs_attestation_protocol.md:
```diff
@@ -274,3 +274,126 @@ We define the error return payload as follows:
+   the response, set to a [KBS Challenge](#challenge) JSON payload.
+2. `/kbs/v0/attest` only accepts `POST` requests which body is a [KBS Attestation](#attestation)
+   JSON payload and the header contains a `Cookie` set to the value received in
+   step 2.a. This is how the attester replies to attestation challenge received
```
Thanks, fixed.

sameo commented on docs/kbs_attestation_protocol.md:
```diff
@@ -274,3 +274,126 @@ We define the error return payload as follows:
+attestation results.
+
+To request a protected resource from the KBS, the attester sends a `GET` request
+to a resource specific endpoint. If the attester is allowed to access the
```
Yes, I added that section now.
Note that I'm planning to also add an OpenAPI definition to this PR. Stay tuned ;-)

sameo commented:
@slp @jialez0 @jiazhang0 The document update is ready to review. I'm planning to add a OpenAPI yaml document to that PR, it would be great to get your review on the HTTP integration part first.

jodh-intel commented on docs/kbs_attestation_protocol.md:
```diff
@@ -108,113 +146,303 @@ attestation challenge to KBC. The payload format is as follows:
+         * The nonce in the Challenge; its hash needs to be included in HW-TEE
+         * evidence and signed by HW-TEE hardware.
+         */
         "nonce": "",
```
Is it worth changing lines like this to something like:

```suggestion
        "nonce": "$nonce",
```

... since currently if you don't read the comments and only look at the JSON, it suggests this field is / should be blank.

sameo replied:
Fixed, thanks.

jodh-intel commented on docs/kbs_attestation_protocol.md:
```diff
@@ -108,113 +146,303 @@ attestation challenge to KBC. The payload format is as follows:
-(Note that the `/*...*/` comments are not valid in JSON web token, and must not be used in real token.)
+The authentication service is provided by the KBS through two endpoints:
+
+1. `/kbs/v0/auth` only accepts `POST` requests which body is a [KBS Request](#request)
```
```suggestion
1. `/kbs/v0/auth` only accepts `POST` requests whose body is a [KBS Request](#request)
```

sameo replied:
Fixed, thanks.

jodh-intel commented on docs/kbs_attestation_protocol.md:
```diff
@@ -108,113 +146,303 @@ attestation challenge to KBC. The payload format is as follows:
+   the attester and its attestation results with that cookie.
+   b. An attestation challenge for the attester to take. This is the content of
+   the response, set to a [KBS Challenge](#challenge) JSON payload.
+2. `/kbs/v0/attest` only accepts `POST` requests which body is a [KBS Attestation](#attestation)
```
```suggestion
2. `/kbs/v0/attest` only accepts `POST` requests whose body is a [KBS Attestation](#attestation)
```

sameo replied:
Fixed, thanks.

jodh-intel commented on docs/kbs_attestation_protocol.md:
```diff
@@ -108,113 +146,303 @@ attestation challenge to KBC. The payload format is as follows:
+     ┃                 ┌─────────────────┐               ┃
+     ┃─────────────────┼POST /kbs/v0/auth┼──────────────▶┃
+     ┃                 ├─────────────────┤               ┃
+     ┃                 │Header: N/A      │               ┃
```
When you say "n/a" do you mean literally empty, or just nothing of interest to show here?
For example, are `Content-Length` and `Content-Type` required? What about a date? If specified, is that validated (what if the date was the epoch or a value far in the future)?

sameo replied:
I mean that the header should not contain any kbs related cookie (as opposed  to the next ones).
I tried to fix it now, let me know if that looks better.

jodh-intel commented on docs/kbs_attestation_protocol.md:
```diff
@@ -108,113 +146,303 @@ attestation challenge to KBC. The payload format is as follows:
+│ Attester │                                          │ KBS │
+└────┳─────┘                                          └──┳──┘
+     ┃                 ┌─────────────────┐               ┃
+     ┃─────────────────┼POST /kbs/v0/auth┼──────────────▶┃
```
Micro-nit: The horizontal lines with arrow heads don't seem to render as designed as they seem to overflow slightly, making the vertical lines bulge. Maybe you need to insert some sort of unicode "half space" char on the other lines to make them line up?

sameo replied:
Fixed it, thanks.

jodh-intel commented on docs/kbs_attestation_protocol.md:
````diff
@@ -108,113 +146,303 @@ attestation challenge to KBC. The payload format is as follows:
+   not.
+
+```
+┌──────────┐                                          ┌─────┐
````
I :heart:  these diagrams! :smile:

sameo replied:
https://monodraw.helftone.com/#introduction ;-)

jodh-intel commented on docs/kbs_attestation_protocol.md:
```diff
@@ -108,113 +146,303 @@ attestation challenge to KBC. The payload format is as follows:
+    ┃                    ┌────────┐                     ┃
+    ┃◀───────────────────┼Response┼─────────────────────┃
+    ┃     ┌──────────────┴────────┴───────────────┐     ┃
+    ┃     │Header: "Set-Cookie kbs-session-id=123"│     ┃
```
How does the KBS expire these session IDs to avoid having to cache them forever (aka a potential slow memory leak)? There doesn't appear to be a final call that the Attester makes. But even if there was that wouldn't be sufficient to deal with a buggy / nefarious attester. I'm assuming some sort of timeout, but where would that need to be specified / documented?

slp replied:
I think it's reasonable leaving that up to the implementation and out of the scope of the protocol. Managing the lifetime of the sessions is a extremely common thing in web development and there is plenty of literature about best practices.

jodh-intel replied:
Maybe we could note explicitly that this is out of scope to avoid any confusion then? I would imagine that a very long expiration time could be problematic from a memory and security standpoint.

jodh-intel commented on docs/kbs_attestation_protocol.md:
```diff
@@ -244,50 +475,21 @@ The [JSON web token](https://jwt.io/) standard is adopted for the token. The for
-3. KBC sends `GET /kbs/key/key_id` request to KBS again, but the payload of this request is the `Attestation` message in the "RCAR" semantics.
-4. KBS returns an `HTTP/1.1 200 OK` response to KBC and the payload is the `Response` message in the "RCAR" semantics.
+The payload part is divided into official claim set and custom claim set. In the
+official claim set, we select three fields, `exp`, `iat`' and `iss`, which
+respectively declare the expiration time, issuing time and issuer (KBS URL
+address) of the token. The custom field in the payload contains the HW-TEE's
```
I wonder if a table would make this clearer:

| field | Description | Type | Example |
|-|-|-|-|
| `exp` | Expiration time | seconds and nanoseconds since epoch? | ? |
| `iat` | Issuing time | seconds and nanoseconds since epoch? | ? |
| `iss` | Issuer | KBS URL | ? |

Aside: Are these fields validated? And if so how? I've mentioned this before in various other places, but how are the timestamps validated, for example, to ensure they have not been spoofed? Does the overally system rely on Secure NTP for example?

jodh-intel commented on docs/kbs_attestation_protocol.md:
```diff
@@ -244,50 +475,21 @@ The [JSON web token](https://jwt.io/) standard is adopted for the token. The for
+respectively declare the expiration time, issuing time and issuer (KBS URL
+address) of the token. The custom field in the payload contains the HW-TEE's
+public key sent by KBC along with the attestation evidence, which is valid
+within the validity period of the token. When KBC uses the token to request
```
What is a "reasonable" or normal validity period?

jodh-intel reviewed (commented):
Thanks @sameo - a few initial comments...

slp commented:
@sameo Thanks a lot for working on this. A couple of comments about the payloads:

- I think it'd be better to drop the inner object and just put the fields on the root object, as it makes parsing the payloads simpler and cleaner. As an example, `Request` would change from this:

```
{
	"request": {
		"version": "0.1.0",
		"tee": "",
		"extra-params": {}
	}
}
```

To this:

```
{
	"version": "0.1.0",
	"tee": "",
	"extra-params": {}
}
```

- `Request` needs a field to identify the workload running in the TEE. In `kbs-types`/`reference-kbs` I've used `workload_id`, but feel free to use the name you think fits better here. This field is needed to tell the Attestation Server against which stored launch measurement should try to match the contents of `tee_evidence` from the `Attestation` payload. This is convenient not only to avoid the need to check against all registered launch measurements, but also to enable per-workload resource unlocking (and some other tricks, such as one time tickets for workloads), as it's likely multiple workloads will share the same launch measurement (same number of vCPUs + same guest initial payload).

jialez0 commented:
> * `Request` needs a field to identify the workload running in the TEE. In `kbs-types`/`reference-kbs` I've used `workload_id`, but feel free to use the name you think fits better here. This field is needed to tell the Attestation Server against which stored launch measurement should try to match the contents of `tee_evidence` from the `Attestation` payload. This is convenient not only to avoid the need to check against all registered launch measurements, but also to enable per-workload resource unlocking (and some other tricks, such as one time tickets for workloads), as it's likely multiple workloads will share the same launch measurement (same number of vCPUs + same guest initial payload).

@slp Does workload here refer to the user's container? I'm not sure whether our remote attestation system will provide services for applications in the user's workload in the future. At least so far, AA's functional scope is only to provide services for Kata components in Pod, including image-rs and ocicrypt-rs in Kata-agent. Therefore, in fact, our remote attestation system is pod granular. When the pod needs a key to decrypt the container image, or needs some confidential resources, such as the pod's security policy configuration file, it will send a request to KBS.

Or is the workload ID you mean here actually the pod ID? If so, how can we identify a pod in the k8s cluster?

slp commented:
> @slp Does workload here refer to the user's container? I'm not sure whether our remote attestation system will provide services for applications in the user's container in the future. At least so far, AA's functional scope is only to provide services for components in Pod, such as image-rs and ocicrypt-rs in Kata-agent. Therefore, in fact, our remote attestation system is pod granular. When the pod needs a key to decrypt the container image, or needs some confidential resources, such as the pod's security policy configuration file, it will send a request to KBS.
> 
> Or is the workload ID you mean here actually the pod ID? If so, how can we identify a pod in the k8s cluster?

@jialez0 The definition of `workload` varies from one use case to the other, and it's more related to the process in which a launch measurement is registered (which is not covered by the KBS protocol) than the Attestation Server itself. Ultimately, it's just a way to know which launch measurement should be used to attest a TEE.

I don't know much about Kata's specifics, but there are some challenges that apply to every Confidential Computing deployment model. At some point, before launching the TEE, something needs to generate the launch measurement for that particular TEE, based on the initial guest payload and the number of vCPUs, and register it with the Attestation Server. And we need a way to relate the launch measurement with the TEE, hence the need for `workload_id`.

sameo commented on docs/kbs_attestation_protocol.md:
```diff
@@ -1,99 +1,137 @@
-request resources or services from KBS by virtue of the token without going through step 2 and step 3.
-
-## Request
+The purpose of the attestation between KBS and KBC is actually to confirm
```
Fixed, thanks.

sameo commented on docs/kbs_attestation_protocol.md:
```diff
@@ -1,99 +1,137 @@
+of this document scope.
+In addition, it should be noted that the confidentiality protection provided by
+HTTPS alone is not enough to meet the security requirements of KBS protocol,
+therefore, as mentioned above, the key generated by HW-TEE needs to be used to
```
Fixed, thanks.

sameo commented on docs/kbs_attestation_protocol.md:
```diff
@@ -1,99 +1,137 @@
-2. Left blank, which means the KBC is implictly expecting to go through attestation 
-and KBS will reply with a Challenge.
+Used to declare the type of HW-TEE platform where KBC is located, the valid
+values are `tdx`, `sgx` and `sev-snp`.
```
It's not too late, and your suggestion makes sense. Thanks.

sameo commented on docs/kbs_attestation_protocol.md:
```diff
@@ -1,99 +1,137 @@
-for example, some attestations depend on Diffie–Hellman to build a secure channel and transfer secret messages 
-(Such as the SEV(-ES) pre-attestation).
+In the run-time attestation scenario (TDX, SEV-SNP, SGX), the `extra-params` are
+not used, so it is kept empty. However, for the attestation of some special
```
Fixed, thanks.

sameo commented on docs/kbs_attestation_protocol.md:
```diff
@@ -1,99 +1,137 @@
-attestation challenge to KBC. The payload format is as follows:
+If the KBC does not own any KBS generated HTTP Cookie, or if the Cookie validity
+has expired, the KBC is not allowed to directly request resources. It must go
+through the attestation process and sends an initial `Request` payload to the
```
Fixed, thanks.

slp commented:
Thanks @sameo, a couple of comments:

- The `Response` payload still has a nested object (thanks for dropping it from the other payloads).
- While re-reviewing the payloads, I've noticed that the fields `nonce` and `tee` from `Attestation` are probably redundant. The first one is provided by the server to the client in `Challenge`, while the second is provided by the client to the guest in `Request`. This doesn't do much harm, but it's a bit confusing since it may look like they can be different from the ones exchanged before.

sameo commented on docs/kbs_attestation_protocol.md:
```diff
@@ -1,99 +1,137 @@
-"Request-Challenge-Attestation-Response" (RCAR) method to facilitate guest attestation and secret injection.
+The Key Broker Service attestation protocol defines communication between a
+Key Broker Client (KBC) in a confidential guest and a trusted Key Broker
+Service  (KBS). The protocol uses the simple, universal, and extensible
```
Fixed, thanks.

sameo commented on docs/kbs_attestation_protocol.md:
```diff
@@ -1,99 +1,137 @@
+   Cookie generated by the KBS during step 2, the KBC can directly
+   request resources or services from KBS, without going through steps 2 and 3.
+
+
```
Fixed, thanks.

sameo commented on docs/kbs_attestation_protocol.md:
```diff
@@ -108,113 +146,303 @@ attestation challenge to KBC. The payload format is as follows:
-The fresh number passed to KBC. KBC needs to place it in the evidence sent to KBS in the next 
-step to prevent replay attacks.
+The fresh number passed to KBC. KBC needs to place it in the evidence sent to
+the KBS in the next step to prevent replay attacks.
```
Not to my knowledge, no.

sameo commented:
Thanks @slp for the review.

> Thanks @sameo, a couple of comments:
> 
>     * The `Response` payload still has a nested object (thanks for dropping it from the other payloads).

Sorry, fixed now.

>     * While re-reviewing the payloads, I've noticed that the fields `nonce` and `tee` from `Attestation` are probably redundant. The first one is provided by the server to the client in `Challenge`, while the second is provided by the client to the guest in `Request`. This doesn't do much harm, but it's a bit confusing since it may look like they can be different from the ones exchanged before.

I agree it could be confusing. I removed them.

Before I add the OpenAPI description to that PR, we need to resolve the open around the `workload_id` proposal. From what I see from the `kbs-reference` implementation, something registers a workload against the attestation service and maps its launch measurement to a workload ID. Is that correct?
I propose to make that field optional, as  other attestation flows may run the evidence verification from some of the attestation parameters like e.g. the guest parameters (SVN, firmware versions, number of vcpus, etc).

slp commented:
> Before I add the OpenAPI description to that PR, we need to resolve the open around the `workload_id` proposal. From what I see from the `kbs-reference` implementation, something registers a workload against the attestation service and maps its launch measurement to a workload ID. Is that correct?

That's correct. The `/register_workload` router isn't part of the protocol, but it comes very handy for testing the attestation workflow.

> I propose to make that field optional, as other attestation flows may run the evidence verification from some of the attestation parameters like e.g. the guest parameters (SVN, firmware versions, number of vcpus, etc).

The field `workload_id` is basically a way of synthesizing those guest parameters (SVN, firmware versions, number of vcpus...) into a single string, which can be a random UUIDv4 (or a formatted string with the parameters, or something else). Otherwise, the protocol would need to have an object that could hold a potentially arbitrary number of guest parameters.

I'm fine with making it optional, but I think it's worth discussing how the protocol would be used when it's missing to make it more solid.

jialez0 commented:
> * While re-reviewing the payloads, I've noticed that the fields `nonce` and `tee` from `Attestation` are probably redundant. The first one is provided by the server to the client in `Challenge`, while the second is provided by the client to the guest in `Request`. This doesn't do much harm, but it's a bit confusing since it may look like they can be different from the ones exchanged before.

@slp @sameo I agree these two fields seems a little redundant here and remove them can make our kbs protocol more concise. But if we do this removing, let us remember that AS needs `nonce` and `tee` fileds to verify the Evidence, so KBS should manage these two fields and pass them together with `tee-pubkey` and `tee-evidence` to AS in a new structure.

jialez0 commented on docs/kbs_attestation_protocol.md:
```diff
@@ -1,220 +1,402 @@
+Keys are generic protected resources that an authenticated attester can get by
+sending a `GET` HTTP request to the `/kbs/v0/key/<key_id>` endpoint.
+
+The `<key_id>` resource name is defined by the attestation service. There should
```
There may be some confusion here. The attestation service should only be responsible for the verification of the evidence in the `Attestation` step, and the resource name should be defined by KBS itself.

sameo replied:
Thanks, I'll fix this as this could be quite confusing.

sameo replied:
Fixed, thanks.

sameo commented on docs/kbs_attestation_protocol.md:
```diff
@@ -1,220 +1,402 @@
-    }
+    /*
+     * The output of the KBS service API.
+     * It must be encrypted with the ephemeral symmetric key generated by the KBC.
```
Thanks, I already fixed that one. I am going to send and updated version soon.

sameo replied:
Fixed now.

sameo commented:
> > Before I add the OpenAPI description to that PR, we need to resolve the open around the `workload_id` proposal. From what I see from the `kbs-reference` implementation, something registers a workload against the attestation service and maps its launch measurement to a workload ID. Is that correct?
> 
> That's correct. The `/register_workload` router isn't part of the protocol, but it comes very handy for testing the attestation workflow.
> 
> > I propose to make that field optional, as other attestation flows may run the evidence verification from some of the attestation parameters like e.g. the guest parameters (SVN, firmware versions, number of vcpus, etc).
> 
> The field `workload_id` is basically a way of synthesizing those guest parameters (SVN, firmware versions, number of vcpus...) into a single string, which can be a random UUIDv4 (or a formatted string with the parameters, or something else). Otherwise, the protocol would need to have an object that could hold a potentially arbitrary number of guest parameters.
> 
> I'm fine with making it optional, but I think it's worth discussing how the protocol would be used when it's missing to make it more solid.

Don't get me wrong, I like the idea and I think it's a simple way to bind a set of measurements/reference values to a workload or a set of workloads. But how is that different from hashing the "unstable" parts  of the attestation evidence, where "unstable" here means everything that is workload specific (code measurement, launch params, num of vcpus, SVNs, etc)? That hash would represent the workload itself, and could otoh be generated by the attestation service.

Adding a `workload_id` to the `Request` payload (even if it's optional) means we need to define how the guest must generate it, and in all cases it implies some a-priori knowledge from the guest about how the measurements were generated and registered in the attestation service. You don't want to add a `workload_id` that wouldn't match exactly how it was registered against a set of measurements that were generated when e.g. creating the guest artifacts.
When not including a `workload_id`, the guest does not need that a-priori knowledge and it is up to the verifier to match the evidence against a set of measurement (And this could be done for example by hashing a subset of the attestation evidence, when receiving it on the attestation service side).

sameo commented:
@jialez0 @Xynnn007 @slp @jodh-intel  I updated the document, addressing a bunch of comments from @jodh-intel and a few from others.

One noticeable change is the proposal to use JSON Web standards for our public key formats (JWK), our Response payload (JWE) and our additional error information (a.k.a. `Problem Details for HTTP APIs`), as suggested by @thomas-fossati. Please have a look and let us know if it makes sense.

jialez0 reviewed (approved):
LGTM. Now our protocol has finally become concise, standardized and beautiful, I like it! Thanks a lot @sameo .

In addition, I have some new views on the proposal of workload ID, but I think it should go beyond the scope of this PR. we can continue to discuss it in #7 .

slp commented:
> In addition, I have some new views on the proposal of workload ID, but I think it may go beyond the scope of this PR. we can continue to discuss it in #7 . @slp

@jialez0 Your proposal sounds interesting, but shouldn't the `measurements` object be part of the KBS protocol?

jialez0 commented:
> @jialez0 Your proposal sounds interesting, but shouldn't the `measurements` object be part of the KBS protocol?

@slp If we put the `measurement` field in the KBS protocol, it can be placed in the Evidence payload, as follows:

```json
{
  "tee-pubkey":"$tee-pubkey",
  "measurements": {},
  "tee-evidence":"$tee-evidence"
}
```

But at the same time, it also means that on the attester side, we need to have a general mechanism for generating and obtaining measurement values, but I am not sure whether different HW-TEE types have different mechanisms for generating and recording measurement values (for example, as far as I know, for Intel TDX, the generation of kernel measurement value is completed by virtual firmware and recorded in the Eventlog under the TCG standard). However, for me personally, it seems great to have a common mechanism for generating and obtaining metrics.

What do you think? @sameo

jodh-intel commented on docs/kbs_attestation_protocol.md:
```diff
@@ -1,220 +1,418 @@
+asymmetric key generated by HW-TEE is used to encrypt the response output
+payload, and the standard [HTTP Cookie](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies)
+mechanism is used to avoid the performance problems caused by multiple
+attestation: it's fairly typical to request for several resources, and then send
+from the same KBC multiple Requests. Having to do the whole attestation dance
```
This might be slightly clearer?:

```suggestion
attestation: it's fairly typical for the KBC to require several resources, resulting in multiple Requests to the KBS. Having to do the whole attestation dance
```

sameo replied:
Fixed, thanks.

jialez0 commented:
> But at the same time, it also means that on the attester side, we need to have a general mechanism for generating and obtaining measurement values, but I am not sure whether different HW-TEE types have different mechanisms for generating and recording measurement values (for example, as far as I know, for Intel TDX, the generation of kernel measurement value is completed by virtual firmware and recorded in the Eventlog under the TCG standard). However, for me personally, it seems great to have a common mechanism for generating and obtaining measurement values.

Or, we can add an interface to each HW-TEE's attester diver to obtain measurements, so that each HW-TEE's attester diver has two interfaces, one for obtaining `tee-evidence`, which is often called report or quote; the other is used to obtain measurements. That sounds like a good idea, too.

jodh-intel commented on docs/kbs_attestation_protocol.md:
```diff
@@ -1,220 +1,418 @@
+   is safe and reliable. The KBC sends an `Attestation` payload to the KBS, that
+   contains the attestation evidence and the HW-TEE generated, ephemeral public
+   key.
+4. **Response**: Upon succesfull attestation, the KBC can request resources by
+   sending HTTP `GET` requests to resource specific endpoints. The KBS responds
```
Could you change the order of these sentences to make this step clearer? I think it would be clearer to state first that the KBS will return a successful `Response` payload iff the KBC's `Attestation` payload was successfully validated.

sameo replied:
I tried to make it clearer now. Let me know if it's not.

jodh-intel commented on docs/kbs_attestation_protocol.md:
````diff
@@ -1,220 +1,418 @@
+    "extra-params": {}
 }
 ```
+<sub>Note that the `/*...*/` comments are not valid in JSON, and must not be
````
You could use a ["note" as we do in Kata](https://github.com/kata-containers/kata-containers/blob/main/docs/Documentation-Requirements.md#notes) to make this stand out (and avoid the need for embedded HTML tags). Same comment for the other instances in this doc.

sameo replied:
Nice, suggestion applied.

jodh-intel commented on docs/kbs_attestation_protocol.md:
```diff
@@ -1,220 +1,418 @@
-If other requests have been made before this request, in the previous response KBS would return a 
-token to KBC. There are two cases for the contents of this field: 
+If the KBC does not own any KBS generated HTTP Cookie, or if the Cookie validity
+has expired, the KBC is not allowed to directly request resources. It must go
```
> the KBC is not allowed to directly request resources.

Is this worth clarifying the meaning of "allowed" here? Presumably if the KBC attempts to use an expired cookie, the KBS will return an error?

sameo replied:
Yes. Clarified now.

jodh-intel commented on docs/kbs_attestation_protocol.md:
```diff
@@ -1,220 +1,418 @@
+receives an attestation challenge from the KBS.
 
-- `extra-params`
+The `tee-pubkey` follows the  [JWK](https://www.rfc-editor.org/rfc/rfc7517)
```
```suggestion
The `tee-pubkey` follows the  [JSON Web Key (JWK)](https://www.rfc-editor.org/rfc/rfc7517)
```

sameo replied:
Fixed, thanks.

jodh-intel commented on docs/kbs_attestation_protocol.md:
```diff
@@ -1,220 +1,418 @@
-- `output`
+# HTTP Integration
+
+KBS uses the HTTP transport protocol to exchange the above described
```
This is rather confusing as the introduction says we use HTTPS (line 27).

sameo replied:
Fixed, thanks.

jodh-intel commented on docs/kbs_attestation_protocol.md:
```diff
@@ -1,220 +1,418 @@
+would be time consuming and have a latency impact on the guest.
+
+In order to ensure the ease of use and security completeness of KBS, we will use
+[HTTPS](#http-integration) as the default transport protocol to carry the
```
> the default transport protocol

What are the alternatives to the HTTP default? How would they be configured?

sameo replied:
There are no alternatives :)

jodh-intel commented on docs/kbs_attestation_protocol.md:
```diff
@@ -1,220 +1,418 @@
+│ Attester │                                         │ KBS │
+└───┳──────┘                                         └──┳──┘
+    ┃            ┌─────────────────────────┐            ┃
+    ┃────────────│GET /kbs/v0/<resource_id>│───────────▶┃
```
What if `<resource_id>`  is `auth` or `attest`? Could that be confusing to clients or KBS programmers, and hence potentially dangerous from a security pov? I see that this endpoint requires `GET` and those "privileged" commands are using `POST`, but would it be safer if this endpoint was different to the "privileged" one (maybe `/kbs/v0/request/<resource_id>` or siimilar?)

sameo replied:
I added a `resources/` prefix. And it can either be `resources/key/keyid` or `resources/token/`

jodh-intel commented on docs/kbs_attestation_protocol.md:
```diff
@@ -1,220 +1,418 @@
-the token. If there is already a token in the KBC's request, no new token will be issued and this 
-field will be left blank.
+Authenticated attesters can also request a token from the KBS, from the
+`/kbs/v0/token` endpoint. Attesters can use tokens to request additional
```
Is this accessible via `GET`?

jodh-intel reviewed (commented):
Thanks @sameo - a few comments...

sameo commented on docs/kbs_attestation_protocol.md:
```diff
@@ -1,220 +1,418 @@
-## Request
+The purpose of the attestation between KBS and KBC is to confirm whether **the
+platform where the KBC is located is in the expected security state**, i.e. if
+it runs in a real and harmless HW-TEE, and the software stack is measured and
```
I meant trustworthy. Fixed now, thanks.

sameo commented on docs/kbs_attestation_protocol.md:
```diff
@@ -1,220 +1,418 @@
+it runs in a real and harmless HW-TEE, and the software stack is measured and
+verifiable.
+
+In this document, HW-TEE attestation process is the semantics of the application
```
Fixed, thanks.

sameo commented on docs/kbs_attestation_protocol.md:
```diff
@@ -1,220 +1,418 @@
+Therefore, as mentioned above, the key generated by HW-TEE needs to be used to
+encrypt and protect the confidential data.
+
+# `RCAR` semantics
```
Fixed, thanks.

sameo commented on docs/kbs_attestation_protocol.md:
```diff
@@ -1,220 +1,418 @@
+
+In this document, HW-TEE attestation process is the semantics of the application
+layer, which is defined as a simple, universal and extensible
+"Request-Challenge-Attestation-Response" (RACR) protocol. The temporary
```
Fixed, thanks.

thomas-fossati commented on docs/kbs.yaml:
```diff
@@ -0,0 +1,267 @@
+                name: kbs-session-id
+                schema:
+                  type: string
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ResponseHeader'
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ResponsePayload'
```
This looks like it's defining two response bodies.  I don't think this is what you mean :-)  Also, syntactically, I guess this would not validate against the OpenAPI schema.  BTW, have you tried this against one of the [validation tools](https://openapi.tools/#description-validators)?  As a recommendation, I'd hook one in (e.g., via Makefile) to check both as you develop and as part of the CI.

sameo replied:
@thomas-fossati I wonder what would be the proper way to represent a JWT or a JWE with an OpenAPI schema.

Using the Compact Serialization scheme, iiuc it would be something like:

```yaml
Response:
     required:
          - jwe
        type: object
        properties:
          jwe: string
          description: A JWE serialized with Compact Serialization
```

A less compact option would be to use the Flattened Serialization syntax, which would make it look like that:

```yaml
Response:
      required:
        - unprotected
        - ciphertext
        - encrypted_key
      type: object
      properties:
        unprotected:
          type: string
        ciphertext:
          type: string
        encrypted_key:
          type: string
      description: JWE serialized with the Flattened Serialization format
```

AFAICT there are no OpenAPI standard ways of defining a JWE or JWS.

sameo replied:
@thomas-fossati Thanks for the offline advice (https://json-schema.org/draft/2020-12/json-schema-validation.html#section-8.6).
I think the OpenAPI 3.1 description looks good now, and it's validated with the redocly linter.

thomas-fossati replied:
👍 awesome job!

thomas-fossati commented on docs/kbs.yaml:
```diff
@@ -0,0 +1,267 @@
+                name: kbs-session-id
+                schema:
+                  type: string
+            application/json:
+              schema:
+                $ref: '#/components/schemas/TokenHeader'
+            application/json:
+              schema:
+                $ref: '#/components/schemas/TokenPayload'
```
ditto

thomas-fossati reviewed (commented):
The protocol flow looks good.  I am not sure the openapi spec is fully correct though.

jodh-intel commented on docs/kbs.yaml:
```diff
@@ -0,0 +1,267 @@
+openapi: 3.0.1
+info:
+  title: Key Broker Service (KBS) API
+  description: HTTP based API for describing the communication between a a
```
```suggestion
  description: HTTP based API for describing the communication between a
```

sameo replied:
Fixed, thanks.

jodh-intel commented on docs/kbs.yaml:
```diff
@@ -0,0 +1,267 @@
+  license:
+    name: Apache 2.0
+    url: http://www.apache.org/licenses/LICENSE-2.0.html
+  version: 0.1.0
```
I wonder if it's worth moving `version` up above `license:` to avoid any confusion with it being `license.version` ;)

jodh-intel commented on docs/kbs.yaml:
```diff
@@ -0,0 +1,267 @@
+info:
+  title: Key Broker Service (KBS) API
+  description: HTTP based API for describing the communication between a a
+               Key Broker Client in a confidential guest and a trusted
```
For parity with KBS:

```suggestion
               Key Broker Client (KBC) in a confidential guest and a trusted
```

sameo replied:
Fixed as well, thanks.

jodh-intel commented on docs/kbs.yaml:
```diff
@@ -0,0 +1,267 @@
+        - enc
+      type: object
+      properties:
+        alg:
+          type: string
+        enc:
+          type: string
```
Does OpenAPI allow you to add a description for `alg` and `enc` as they are brutally terse ;)

jodh-intel commented on docs/kbs.yaml:
```diff
@@ -0,0 +1,267 @@
+      properties:
+        ciphertext:
+          type: string
+        iv:
```
What is this? Can it also have a description?

sameo replied:
Description fixed now.

jodh-intel commented on docs/kbs.yaml:
```diff
@@ -0,0 +1,267 @@
+
+    TokenPayload:
+      required:
+        - ext
+        - iat
+        - iss
+        - tee_pubkey
```
Could these be documented too?

sameo replied:
They are documented now.

jodh-intel commented on docs/kbs.yaml:
```diff
@@ -0,0 +1,267 @@
+
+    PublicKey:
+      required:
+        - kty
+        - alg
+        - k
```
Ditto ;)

sameo commented:
@jodh-intel Thanks a bunch for the thorough review. I think most of your comments are addressed now.

jodh-intel commented on docs/kbs.yaml:
```diff
@@ -0,0 +1,297 @@
+          type: string
+        tee:
+          type: string
+          enum: [intel-tdx, intel-sgx, amd-sev]
```
Nit: Unsorted list.

sameo replied:
Fixed, thanks.

sameo commented on docs/kbs.yaml:
```diff
@@ -0,0 +1,297 @@
+          description: Key type
+        alg:
+          type: string
+          description: Key Algorithn
```
Fixed as well.

sameo commented on docs/kbs.yaml:
```diff
@@ -0,0 +1,297 @@
+          description: The encrypted key used for encrypting the response data.
+        if:
+          type: string
+          description: Encryption algorithm initialisation vector.
```
Good catch, fixed now.

sameo commented:
@slp @jialez0 I think we're good to go with that PR. Could you please have one last look?

slp commented:
@sameo If you don't mind waiting a bit, I'd like to update kbs-types/reference-kbs to see how the protocol looks like. I'll try to find some time to do it this week.

slp commented on docs/kbs_attestation_protocol.md:
````diff
@@ -1,220 +1,412 @@
+     */
+    "encrypted_key": "$enc_kbs_symkey"
 }
 ```
````
JSON lint tells me this isn't valid JSON. Is this intended to be an array? Can the JOSE header object be a field within the root object?

sameo replied:
@slp Yes, the example is incorrectly formatted and the intention is for it to be an array, yes. So it should really look like that:

```json
[{
	"alg": "$alg",
	"enc": "$enc"
}, {
	"ciphertext": "$response_output",
	"iv": "$initialization_vector",
	"encrypted_key": "$enc_kbs_symkey"
}]
```

slp replied:
I'm not familiar with the JOSE specs. Is it required to be found in this form? As ordering in an array isn't guaranteed, consuming this payload would require to peek into the object to find out its format.

Would it be possible for the JOSE header to be an object in named field? Something like:

```json
{
	"ciphertext": "$response_output",
	"iv": "$initialization_vector",
	"encrypted_key": "$enc_kbs_symkey",
	"jose_params": {
		"alg": "$alg",
		"enc": "$enc"
	}
}
```

Xynnn007 replied:
> I'm not familiar with the JOSE specs. Is it required to be found in this form? As ordering in an array isn't guaranteed, consuming this payload would require to peek into the object to find out its format.
> 
> Would it be possible for the JOSE header to be an object in named field? Something like:
> 
> ```json
> {
> 	"ciphertext": "$response_output",
> 	"iv": "$initialization_vector",
> 	"encrypted_key": "$enc_kbs_symkey",
> 	"jose_params": {
> 		"alg": "$alg",
> 		"enc": "$enc"
> 	}
> }
> ```

AFAIK, the architecture of a JOSE object should look like
```rust
BASE64URL_ENCODE({"alg": "$alg","enc": "$enc"}).
BASE64URL_ENCODE($enc_kbs_symkey).
BASE64URL_ENCODE($initialization_vector).
BASE64URL_ENCODE($response_output).
BASE64URL_ENCODE($authentication_tag)
```

Here `$authentication_tag` is extra result of specific symmetric encryption (s.t. AEAD) besides cipher text. Algorithms like `AES/GCM` will produce such authentication data. And care about each adjacent two base64url-encoded part are connected by a dot `.`

I think the json format shown in the document may be for understanding the fields of the JWE used here. Am I right? @sameo 

Reference to [JOSE Cookbook](https://datatracker.ietf.org/doc/html/rfc7520#section-5.1.5) and [JWE Spec](https://www.rfc-editor.org/rfc/rfc7516#section-2).

sameo replied:
Sorry for the delayed answer, post LPC has been busy for me...

@Xynnn007 you're partly right, although what you're describing is the [compact JWE serialization format](https://www.rfc-editor.org/rfc/rfc7516#section-7.1).
But I think we want to use the [flattened JWE JSON serialization format](https://www.rfc-editor.org/rfc/rfc7516#section-7.2.2). 

With that format, our `Response` JSON-formatted payload would then look like that:

```json
{
    "protected": "$jose_header",
    "encrypted_key": "$encrypted_key",
    "iv": "$iv",
    "ciphertext": "$ciphertext",
    "tag": "$tag"
}
```

Where:

```shell
jose_header = BASE64URL_ENCODE({"alg": "$alg","enc": "$enc"})
encrypted_key = BASE64URL_ENCODE($enc_kbs_symkey)
iv = BASE64URL_ENCODE($initialization_vector)
ciphertext = BASE64URL_ENCODE($response_output)

# tag is optional and depend on the encryption algorithm.
tag = BASE64URL_ENCODE($authentication_tag) 
```

I find it a little convoluted, but this really is what the spec defines. FWIW there are some on going, recent work on a rust crate for JWE and JWK: https://github.com/RustCrypto/JOSE

If we're fine with that definition, I'll update that PR accordingly.

sameo replied:
Thanks @slp. I updated the document and the openAPI description as well. I think we *may* be good now :) PTAL.

slp commented on docs/kbs_attestation_protocol.md:
````diff
@@ -234,43 +426,57 @@ The [JSON web token](https://jwt.io/) standard is adopted for the token. The for
+    "tee-pubkey": $pubkey
 }.
 [Signature]
 ```
````
Ditto.

slp replied:
@sameo Just to be sure, the KBS is expected to send the token pre-encrypted, doesn't it? That is, three base64-URL strings (header, payload and signature) separated by dots. Is that right? If so, perhaps the docs should explicitly state that?

sameo replied:
That's correct @slp. I expanded the token section quite a bit to describe this. I also updated the openAPI description accordingly. Please have a look.

sameo commented:
> @sameo If you don't mind waiting a bit, I'd like to update kbs-types/reference-kbs to see how the protocol looks like. I'll try to find some time to do it this week.

Sure thing. I'm actually happy to not merge this PR until you can verify it's implementable from `reference-kbs`.

jialez0 commented on docs/kbs_attestation_protocol.md:
```diff
@@ -1,220 +1,412 @@
+     * be included in the HW-TEE evidence and signed by the HW-TEE hardware.
+     */
+    "tee-pubkey": $pubkey
+
```
As we discussed earlier, perhaps we need to add a `measurements` field (or another synonymous name) to `Attestation` json payload? Just like this:

```json
{
    "tee-pubkey": $pubkey,
    "measurements": {
        $name: $value,
        ...
    },
    "tee-evidence": {}
}
```

jialez0 replied:
cc @slp @sameo

slp reviewed (approved):
I've just updated `kbs-types` and `reference-kbs`, and I think the protocol, in its current state, fits quite well with web frameworks. Thanks @sameo !

sameo commented:
@jiazhang0 Could you PTAL?
